### ISSUE TITLE
npm i streamsaver -> how to use in react ?

### ISSUE THREAD

**Issue by x4080** [jimmywarting/StreamSaver.js#158]
Hi, I tried it in react using 
```
import streamSaver from 'streamsaver'
```

And it shows error : window is not defined

How to use it with npm ?

Thanks

**Comment by jimmywarting** [comment]
React.js or react native?

**Comment by x4080** [comment]
Hi @jimmywarting thanks for quick reply, its react js, my current workaround is using 
```
<script src="https://cdn.jsdelivr.net/npm/streamsaver@2.0.3/StreamSaver.min.js"></script>
```
and put it in root html

**Comment by x4080** [comment]
I just remembered maybe its because of ssr, this was in my next js project, when it is importing, no window object available

**Comment by dreamofi** [comment]
@x4080 : you import it with: `import * as streamsaver from 'streamsaver'`
You can see an example [here](https://codesandbox.io/s/kjrxh?file=/src/index.js):

**Comment by x4080** [comment]
@dreamofi Thanks, I'll try it

**Comment by x4080** [comment]
@dreamofi Thanks it works !

**Comment by x4080** [comment]
@dreamofi No, I make mistake, it works when the link is in the client, but when rendered from server its not working

**Comment by dreamofi** [comment]
@x4080 can you post your webpack configuration file?

**Comment by x4080** [comment]
@dreamofi , I'm using next js and my next.config.js seems not using custom webpack config so it goes like this
```
const withSass = require('@zeit/next-sass');
const withOffline = require('next-offline')
module.exports = withSass(withOffline());
```

Thanks for your help

**Comment by dreamofi** [comment]
@x4080 I have yet used nextjs, but what is the error when you render from the server, and can you post the code where you use streamsaver? Do you add ponyfill?

**Comment by jonsadka** [comment]
@dreamofi you should be able to load the dependencies on `componentDidMount` 

```
  componentDidMount() {
    // HACK: streamsaver references window which is undefined on SSR. Ensure library is only loaded on client
    try {
      this.streamSaver = require('streamsaver');
      if (!this.streamSaver.WritableStream) {
        this.streamSaver.WritableStream = require('web-streams-polyfill/ponyfill').WritableStream;
      }
    } catch (e) {
      console.error(e);
    }
  }
```

@jimmywarting any chance we can shift this logic to the package itself? Ideally we can make a check against `window` here https://github.com/jimmywarting/StreamSaver.js/blob/1bce493b16fd51cf313570445e2b9f96045393cb/StreamSaver.js#L15

**Comment by jimmywarting** [comment]
maybe should use `globalThis` instead?

**Comment by jonsadka** [comment]
@jimmywarting yeah I think that would work!

**Comment by pedro-pedrosa** [comment]
any updates on this?

**Comment by jimmywarting** [comment]
if window is not defined then you probably are using streamsaver in a wrong context like some server side render like next or  in a web worker.

only include streamsaver on the client side (on the main thread)

**Comment by pedro-pedrosa** [comment]
I'm using dynamic imports on my code for now to work around this. was just asking because there was a suggestion of using `globalThis` in place of `window`.

I guess if StreamSaver.js is supposed to be a browser-only module and `window` is supposed always work, then this issue should be closed to avoid confusion.

**Comment by pmourer** [comment]
Dynamic imports don't really seem like the greatest solution. Why can't this just check for the presence of window or use `globalThis`?

**Comment by eligrey** [comment]
Webpack 5 might be able to solve this issue with its new [native Worker support](https://webpack.js.org/blog/2020-10-10-webpack-5-release/#native-worker-support).
